sql: TestRandomSyntaxFunctions: skip function call and fix CREATE STATS use of span after finish

fqazi opened the pull request:
Previously, this test could depending on the concurrency,
could spawn a large number of schema telemetry jobs via
(ctdb_internal.create_sql_schema_telemetry_job).
This could lead to contention that will eventually
cause this test to time out. To address this, this
patch limits calling the telemetry job creation function.

Fixes: #108153

Release note: None

fqazi committed:
sql: TestRandomSyntaxFunctions skip schema telemetry job function

Previously, this test could depending on the concurrency,
could spawn a large number of schema telemetry jobs via
(ctdb_internal.create_sql_schema_telemetry_job).
This could lead to contention that will eventually
cause this test to time out. To address this, this
patch limits calling the telemetry job creation function.

Fixes: #108153

Release note: None

rafiss reviewed (approved):
thanks! will the backport address the 23.1 failure? https://github.com/cockroachdb/cockroach/issues/107928

fqazi commented:
@rafiss I finally was able to repro the failures, getting them about 1 in 1000 runs. So, let me see what else needs to be added here.

fqazi committed:
sql: fix use of span after finish in CREATE STATISTICS

Previously, the code for creating and waiting for the
job inside CREATE STATISTICs had a race condition, where
we could end up using context after a tracing span was
cleaned up on it. To address this, this patch adjusts
the code to wait for a go routine to clean up before,
returning and allowing the span to clean up.

Epic: none

Release note: None

rafiss commented on pkg/sql/create_stats.go:
```diff
@@ -162,25 +161,33 @@ func (n *createStatsNode) startJob(ctx context.Context) error {
-				ctx, "delete-job", nil /* txn */, stmt, jobID,
-			); delErr != nil {
-				log.Warningf(ctx, "failed to delete job: %v", delErr)
+	ctx, cancelFn := context.WithCancel(ctx)
```
hm, have you already thought about whether this needs to use `stopper.RunAsyncTask`? i think maybe it does not, since this context will end up being cancelled during quiescence, but the Queries team would likely know for sure.

assuming it does not need the `stopper`, i think this could be a bit simpler with a `ctxgroup`. here you would do:


```
	g := ctxgroup.WithContext(ctx)
	g.GoCtx(func(ctx context.Context) error {
		err := job.AwaitCompletion(ctx)
		if err != nil {
			if errors.Is(err, stats.ConcurrentCreateStatsError) {
				// Delete the job so users don't see it and get confused by the error.
				const stmt = `DELETE FROM system.jobs WHERE id = $1`
				if _ /* cols */, delErr := n.p.ExecCfg().InternalDB.Executor().Exec(
					ctx, "delete-job", nil /* txn */, stmt, jobID,
				); delErr != nil {
					log.Warningf(ctx, "failed to delete job: %v", delErr)
				}
			}
		}
		return err
	})
	return g, err
```

and you would have no need for the `select`, and we'd return the `ctxgroup` here instead of the error channel.

then there is no need to use a `cancelFunc`, since the `ctxgroup` itself takes care of propagating context cancellation.

rafiss commented on pkg/sql/create_stats.go:
```diff
@@ -104,26 +104,25 @@ type createStatsNode struct {
-		close(n.run.errCh)
-	}()
-	return nil
+	cancelFn, err := n.startJob(params.ctx)
```
as per the comment a bottom of this file, if you use a ctxgroup this would change to 

```
	g, err := n.startJob(params.ctx)
	n.run.group = g
	return err
```

rafiss commented on pkg/sql/create_stats.go:
```diff
@@ -104,26 +104,25 @@ type createStatsNode struct {
 }
 
 func (n *createStatsNode) Next(params runParams) (bool, error) {
+	defer n.run.cancelFn()
```
and this would just simplify to
```
return false, n.run.group.Wait()
```

rafiss commented on pkg/sql/create_stats.go:
```diff
@@ -162,25 +161,33 @@ func (n *createStatsNode) startJob(ctx context.Context) error {
 	}
 	if err := job.Start(ctx); err != nil {
-		return err
+		return nil, err
```
perhaps a question for the Queries team, but why so we want to start the job here, rather than in `Next()`? wouldn't that let us avoid the need for async tasks?

rafiss reviewed (changes requested):
this should get a review from the Queries team.

also, can you write a test that reproduces the problem? or describe some of the manual steps that can repro it?

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/create_stats.go` line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295396141:-NbzalgODcAn-XrbVlKD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

as per the comment a bottom of this file, if you use a ctxgroup this would change to 

```
	g, err := n.startJob(params.ctx)
	n.run.group = g
	return err
```
</blockquote></details>

Done.
___
*[`pkg/sql/create_stats.go` line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295401666:-NbzamHb3B3lb5NuCwz4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L120)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

and this would just simplify to
```
return false, n.run.group.Wait()
```
</blockquote></details>

Done.
___
*[`pkg/sql/create_stats.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295405425:-NbzanRA0eImr4LYOkU5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L167)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

perhaps a question for the Queries team, but why so we want to start the job here, rather than in `Next()`? wouldn't that let us avoid the need for async tasks?
</blockquote></details>

Done.
___
*[`pkg/sql/create_stats.go` line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295392297:-NbzaoBTDqBrvzcGyIcd:b-dt7gjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L169)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, have you already thought about whether this needs to use `stopper.RunAsyncTask`? i think maybe it does not, since this context will end up being cancelled during quiescence, but the Queries team would likely know for sure.

assuming it does not need the `stopper`, i think this could be a bit simpler with a `ctxgroup`. here you would do:


```
	g := ctxgroup.WithContext(ctx)
	g.GoCtx(func(ctx context.Context) error {
		err := job.AwaitCompletion(ctx)
		if err != nil {
			if errors.Is(err, stats.ConcurrentCreateStatsError) {
				// Delete the job so users don't see it and get confused by the error.
				const stmt = `DELETE FROM system.jobs WHERE id = $1`
				if _ /* cols */, delErr := n.p.ExecCfg().InternalDB.Executor().Exec(
					ctx, "delete-job", nil /* txn */, stmt, jobID,
				); delErr != nil {
					log.Warningf(ctx, "failed to delete job: %v", delErr)
				}
			}
		}
		return err
	})
	return g, err
```

and you would have no need for the `select`, and we'd return the `ctxgroup` here instead of the error channel.

then there is no need to use a `cancelFunc`, since the `ctxgroup` itself takes care of propagating context cancellation.
</blockquote></details>

Let me see queries thoughts on if this should use RunAsyncTask


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
this looks good now! and thanks for the test. let's see what the queries team thinks

yuzefovich reviewed (commented):
Thanks for the fix!

> Reviewed 1 of 1 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-:-Nc-9s3bD1P4y1wNPcSU:b-p1kr18)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rafiss)

___
*[`pkg/sql/create_stats.go` line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295396141:-Nc-92r88CinNkdhagDf:b-tcyk8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: we can just do `return n.startJob(params.ctx)`.
___
*[`pkg/sql/create_stats.go` line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295392297:-Nc-9YEqBRLli1EfgT7A:bprbrkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L169)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Let me see queries thoughts on if this should use RunAsyncTask
</blockquote></details>

IIUC we (i.e. DB engineers) do have a slight preference to using the stopper, but using the ctx group here seems fine too, up to you.
___
*[`pkg/sql/create_stats.go` line 16 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-55qIB22oFuAjN3eF:-Nc-55qIB22oFuAjN3eG:b-2k635y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L16)):*
> ```Go
> 	"context"
> 	"fmt"
> 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
> ```

nit: out of order import.
___
*[`pkg/sql/create_stats.go` line 124 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-8rf-9HG5ND-NL0mS:-Nc-8rf-9HG5ND-NL0mT:bqtz22c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L124)):*
> ```Go
> func (*createStatsNode) Values() tree.Datums   { return nil }
> 
> // startJob starts a CreateStats job to plan and execute statistics creation.
> ```

nit: I'd extend the comment to indicate that the job is started synchronously, but another goroutine is started via the ctx group and the caller must wait on that ctx group (unless an error is returned).
___
*[`pkg/sql/stats/create_stats_job_test.go` line 16 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-9OW0BwTR2exnEPlh:-Nc-9OW0BwTR2exnEPli:b-2k635y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L16)):*
> ```Go
> 	"context"
> 	"fmt"
> 	"github.com/stretchr/testify/require"
> ```

nit: out of order import.
___
*[`pkg/sql/stats/create_stats_job_test.go` line 142 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-5Xv71x-ykIB6pOED:-Nc-5Xv71x-ykIB6pOEE:bz9jhcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L142)):*
> ```Go
> 	ctx := context.Background()
> 	const nodes = 1
> 	tc := testcluster.StartTestCluster(t, nodes, params)
> ```

nit: I think `serverutils.StartServer` should be sufficient here (I know it was copied from other tests, but we should be using the server instead of starting 1 node cluster if possible).
___
*[`pkg/sql/stats/create_stats_job_test.go` line 163 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-6EIy8L9kpnzszzWJ:-Nc-6EIy8L9kpnzszzWK:b6tz4tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L163)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		str := sqlDB.QueryStr(t, "SELECT * FROM [SHOW CLUSTER STATEMENTS] ")
> 		fmt.Printf("%s\n", str)
> ```

nit: do we need it? If we do, let's use `t.Log`.
___
*[`pkg/sql/stats/create_stats_job_test.go` line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-6Nxj3o6EnDYvcjKT:-Nc-6Nxj3o6EnDYvcjKU:b-27j3hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L164)):*
> ```Go
> 		str := sqlDB.QueryStr(t, "SELECT * FROM [SHOW CLUSTER STATEMENTS] ")
> 		fmt.Printf("%s\n", str)
> 		row := conn.QueryRow("select query_id from [SHOW CLUSTER STATEMENTS] WHERE query like 'CREATE STATISTICS%';")
> ```

nit: let's capitalize `select`, `from`, and `like` keywords here and below since we're capitalizing some other keywords already.
___
*[`pkg/sql/stats/create_stats_job_test.go` line 172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-64525Qe9YAw9DFCv:-Nc-64525Qe9YAw9DFCw:b-s88rts) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L172)):*
> ```Go
> 		return err
> 	})
> 	allowRequest <- struct{}{}
> ```

Shouldn't this send be before `SucceedsSoon`? The comment above implies that we need to wait until the stats job did one scan, and I don't currently see that.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/create_stats.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-F9pb0lhLtxP-tgme:-Nc-F9pb0lhLtxP-tgmf:b8qa5gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L121)):*
> ```Go
> }
> 
> func (*createStatsNode) Close(context.Context) {}
> ```

do we need to wait for the ctxgroup here or cancel it somehow?

i'm not sure about the contracts of Next and Close, but @yuzefovich would know better.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295405425:-Nc-FcZN1Uu-WMcBc8ND:b-gf1v8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L167)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

what was the answer to this? any thoughts @yuzefovich ? perhaps when adding additional comments, they should also explain why it's done this way


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @rafiss, and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295405425:-Nc-hlgo7TUQR9ng2qcX:b-m759t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L167)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what was the answer to this? any thoughts @yuzefovich ? perhaps when adding additional comments, they should also explain why it's done this way
</blockquote></details>

I'm also curious. Maybe for some queries like `WITH foo AS (CREATE STATISTICS ...) SELECT 1;` we don't ever call `Next()`? Or maybe starting the job from `startExec` gives us the correct ctx?


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295396141:-Nc-zu3v3iWbZ-zUQ-hr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we can just do `return n.startJob(params.ctx)`.
</blockquote></details>

Done.
___
*[`pkg/sql/create_stats.go` line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295392297:-Nc01zdD52yx1SYizM-X:bycgx5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L169)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

IIUC we (i.e. DB engineers) do have a slight preference to using the stopper, but using the ctx group here seems fine too, up to you.
</blockquote></details>

I'll leave it as is for now. I think we would need to go back to having a channel, too I think.
___
*[`pkg/sql/stats/create_stats_job_test.go` line 172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-64525Qe9YAw9DFCv:-Nc03QRt1zREga3tH_4f:b-y3wnon) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/stats/create_stats_job_test.go#L172)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Shouldn't this send be before `SucceedsSoon`? The comment above implies that we need to wait until the stats job did one scan, and I don't currently see that.
</blockquote></details>

Good point. It just needs to wait for the creation of it. So that the asynchronous go routine can be cancelled.

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/create_stats.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295405425:-Nc0CyEG5oximtjPxhDB:b6001kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L167)):*
Hm, I don't see a good reason why we couldn't kick off the job in `Next` (this code was introduced in 9ef8bb27bacb244bcecc183610d080d2b930c0ad when we started using the jobs framework for table stats, and my guess the choice to put "start job" method into "startExec" method seemed more appropriate based on the naming; I also looked over the review comments, and this wasn't discussed; also none of SQL Execution folks (like Jordan and Alfonso) reviewed it).

> Maybe for some queries like `WITH foo AS (CREATE STATISTICS ...) SELECT 1;` we don't ever call `Next()`?

It doesn't seem like it should be possible to construct such a statement (quick look at the grammar suggests that it shouldn't be possible).

> Or maybe starting the job from `startExec` gives us the correct ctx?

`runParams` in `startExec` and `Next` should have the same context (that is set by `planNodeToRowSource` right before it calls `startExec`).

I like Rafi's suggestion, so I think we should move starting the job into `Next` to remove the extra goroutine.
___
*[`pkg/sql/create_stats.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-F9pb0lhLtxP-tgme:-Nc0Bqwa5bvi-an_6fjb:bfozt11) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L121)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do we need to wait for the ctxgroup here or cancel it somehow?

i'm not sure about the contracts of Next and Close, but @yuzefovich would know better.
</blockquote></details>

Yes, good catch. It might not be strictly necessary to wait for the ctx group (because the parent context will be canceled via `FlowBase.mu.ctxCancel` on `Flow.Cleanup`, so it's guaranteed that eventually context of the ctx group is canceled), but I think it'd still be nice to explicitly wait in `Close` for the ctx group (perhaps derive a child context that is canceled explicitly on `Close`).


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/create_stats.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-F9pb0lhLtxP-tgme:-Nc2blaL6HQ184HFw62X:bpcjlc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L121)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yes, good catch. It might not be strictly necessary to wait for the ctx group (because the parent context will be canceled via `FlowBase.mu.ctxCancel` on `Flow.Cleanup`, so it's guaranteed that eventually context of the ctx group is canceled), but I think it'd still be nice to explicitly wait in `Close` for the ctx group (perhaps derive a child context that is canceled explicitly on `Close`).
</blockquote></details>

The issue is that the job wait can return an error, so we need to wait for that in Next. I don't think we are allowed to surface errors in Close.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rafiss, and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108525#gh-1295405425:-Nc2c6O04v0h41pP_eyp:b8i35xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e607b749773fd6da6187b39a95068e764814d8d4/pkg/sql/create_stats.go#L167)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I don't see a good reason why we couldn't kick off the job in `Next` (this code was introduced in 9ef8bb27bacb244bcecc183610d080d2b930c0ad when we started using the jobs framework for table stats, and my guess the choice to put "start job" method into "startExec" method seemed more appropriate based on the naming; I also looked over the review comments, and this wasn't discussed; also none of SQL Execution folks (like Jordan and Alfonso) reviewed it).

> Maybe for some queries like `WITH foo AS (CREATE STATISTICS ...) SELECT 1;` we don't ever call `Next()`?

It doesn't seem like it should be possible to construct such a statement (quick look at the grammar suggests that it shouldn't be possible).

> Or maybe starting the job from `startExec` gives us the correct ctx?

`runParams` in `startExec` and `Next` should have the same context (that is set by `planNodeToRowSource` right before it calls `startExec`).

I like Rafi's suggestion, so I think we should move starting the job into `Next` to remove the extra goroutine.
</blockquote></details>

Done.

Good point, I made it synchronous inside Next
___
*[`pkg/sql/create_stats.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc-F9pb0lhLtxP-tgme:-Nc2cw1JBL4ThrqDfPXv:b-ywgp8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3dc4d6e45975be74ee983f0d718f71b775bdf8a/pkg/sql/create_stats.go#L121)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

The issue is that the job wait can return an error, so we need to wait for that in Next. I don't think we are allowed to surface errors in Close.
</blockquote></details>

Done.

If it is synchronous we don't have to worry about any of this now.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @michae2, and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 106 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sIHZC9lzmc6FGzFM:-Nc2sIHZC9lzmc6FGzFN:b6hd4sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L106)):*
> ```Go
> // execution.
> type createStatsRun struct {
> 	ctxGroup ctxgroup.Group
> ```

we don't need the ctxgroup now
___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc2sPdJ1tUtzeIsKB2Z:b-ul4o02) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
> ```Go
> }
> 
> func (n *createStatsNode) Next(params runParams) (bool, error) {
> ```

this really is a super nit, but would it also work if this were done in `startExec`, and `Next()` was just empty? (i think that's how most of the legacy schema change operations work, so i believe it should be fine, but again the Queries team would know better about the contracts of `startExec` vs `Next`, so let's see what @yuzefovich thinks)
___
*[`pkg/sql/create_stats.go` line 124 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2snGK2H176rUMVSlG:-Nc2snGK2H176rUMVSlH:b-9y2s5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L124)):*
> ```Go
> // statistics creation, and a go routine is started via ctx group,
> // which the caller must wait for afterwards (within createStatsNode).
> func (n *createStatsNode) startJob(ctx context.Context) error {
> ```

nit: maybe now we can rename it to `runJob`


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rafiss, and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 106 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sIHZC9lzmc6FGzFM:-Nc38Xt-0UDPQko09-CD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L106)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we don't need the ctxgroup now
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc3NzOdCGBhdAP7WQkb:by6ohz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this really is a super nit, but would it also work if this were done in `startExec`, and `Next()` was just empty? (i think that's how most of the legacy schema change operations work, so i believe it should be fine, but again the Queries team would know better about the contracts of `startExec` vs `Next`, so let's see what @yuzefovich thinks)
</blockquote></details>

+1 I'd also prefer doing this in `startExec` than `Next` to match the legacy schema change operators.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-:-Nc3QwQzCeyOm5NpmEGv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc3QXWd0yaLyolfnP5V:b3vk4gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

+1 I'd also prefer doing this in `startExec` than `Next` to match the legacy schema change operators.
</blockquote></details>

Hm, my personal preference is to have this job in `Next` because at least in my mind `startExec` should be quick whereas `Next` can block. I don't think it's explicitly called out in the interface contract though, just my own mental model.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc3RK8ZEtyJoo6WOHKz:b-6qitkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, my personal preference is to have this job in `Next` because at least in my mind `startExec` should be quick whereas `Next` can block. I don't think it's explicitly called out in the interface contract though, just my own mental model.
</blockquote></details>

As Rafi points out, legacy create table, alter table, drop table, etc all run synchronously in `startExec` and just have stub `Next` methods. I'm nervous that there's some reason for that.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc3VZcR6uPhIIOYqaan:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

As Rafi points out, legacy create table, alter table, drop table, etc all run synchronously in `startExec` and just have stub `Next` methods. I'm nervous that there's some reason for that.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/create_stats.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-Nc2sPdJ1tUtzeIsKB2Y:-Nc3Y-fEEv2azDZ_U-IR:b8gabyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fcc8d29fb20c7b6096c0cc4c026b956e00afbf/pkg/sql/create_stats.go#L114)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

SGTM for consistency.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108525#-:-Nc3_LbzB7EMAAurJYjt:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for seeing this through!

fqazi commented:
@rafiss @yuzefovich @michae2 TFTR!

bors r+
